### ISSUE TITLE
Serverless Compose?

### ISSUE THREAD

**Issue by mnapoli** [ngneat/nx-serverless#8]
Hi, I stumbled upon this repository and I'm wondering if [Serverless Framework Compose](https://www.serverless.com/blog/serverless-framework-compose-multi-service-deployments) could be useful here?

Furthermore we're exploring possible integrations with Learn/Nx/Turborepo, looking to bounce ideas about that. So I thought I'd open an issue ^^

**Comment by NetanelBasal** [comment]
Hi!

Sure. I'm not very familiar with the new compose capabilities, but Nx (with this setup) already has everything included there and more.

Currently, it's annoying that we need to serve the different services over different ports. I could handle this by creating a proxy server, but it would be great if a solution were integrated into the serverless-offline plugin. 

Feel free to explore the repository, share your ideas, and make suggestions. Thank you for your excellent work creating the serverless framework.

**Comment by mnapoli** [comment]
Indeed Nx takes care of the deployments right now, which is great. What Compose could bring to the table here is the ability to share outputs between services easily.

Instead of hardcoding a resources name (e.g. the DynamoDB table names/ARNs) or using CloudFormation exports/imports (which lock the stacks completely together, which is super painful when you want to change them), you can do:

```yml
# serverless-compose.yml
services:

  core:
    path: services/core

  users:
    path: services/users
    params:
      restApiId: ${core.restApiId}
      tableArn: ${core.tableArn}
  ...
```

This is YAML instead of TS, and very simple, but you get the idea. Compose doesn't use CloudFormation Imports, instead it fetches the CloudFormation output value directly, which is fast and doesn't lock the stacks together.

Also, once you have set up variables between each service, Compose will deploy them in order automatically.

We are looking into improvements that may also interest you:

1. defining parameters in `serverless-compose.yml` that apply to all services (https://github.com/serverless/compose/issues/38):

```yml
# serverless-compose.yml

params:
  restApiId: ${core.restApiId}
  tableArn: ${core.tableArn}

services:

  core:
    path: services/core

  users:
    path: services/users
  ...
```

2. integrating with Nx/Lerna/etc. to deploy only "affected" services (i.e. services with changes)

Oh, also one final advantage to consider with Compose: you can run logs across all functions of all services: https://youtu.be/lPevgIk11Rg?t=783

**Comment by NetanelBasal** [comment]
I'm not sure what the added benefit is. Right now, I can share single source variables within any service. I have the power of Javascript.

Regarding the dependencies, I can tell Nx which service is dependent on which. (https://github.com/ngneat/nx-serverless/blob/main/services/todos/project.json#L60)

**Comment by mnapoli** [comment]
Sorry if I wasn't more clear, let me rephrase the benefit here: Compose would avoid:

- either hardcoding resource names (e.g. DynamoDB table name/ARN, particularly useful for bucket name because they need to be unique/random) https://github.com/ngneat/nx-serverless/blob/81857666fc9f26390160435026d41a15731c579d/environments/environment.prod.ts#L8
- or using CF Export/Import, [which locks stacks together](https://stackoverflow.com/a/64784633/245552) and is very painful to deal with https://github.com/ngneat/nx-serverless/blob/e8afbd13f40614bb18c23a704b18e0fefe999a81/serverless.base.ts#L42-L49

Also my apologies, I forgot to mention that `serverless-compose.yml` can also be written in TS as `serverless-compose.ts`, so it doesn't mean giving up the power of Javascript.

**Comment by NetanelBasal** [comment]
It'll be interesting to see the same project using serverless-compose. Maybe we could do it in a new branch and compare the solutions.
